### PR TITLE
Refactoring App registration

### PR DESCRIPTION
**Pull request by dorjesinpo**
1.  Fix the race when reconfiguring domain in FSM
2. Generate AppKey only once when reconfiguring domain.

### CODE REVIEW

**Review by kaikulimu — CHANGES_REQUESTED** [review]

File: src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp [hunk 1]
```diff
@@ -1880,102 +1858,47 @@ void RootQueueEngine::afterAppIdRegistered(
 
     if (!d_isFanout) {
         BALL_LOG_ERROR << "RootQueueEngine::afterAppIdRegistered() should "
-                       << "never be called for a non-Fanout queue.  Received "
-                       << "call to register appId '" << appIdKeyPair.first
-                       << "', appKey '" << appIdKeyPair.second << "'.";
+                       << "never be called for a non-Fanout queue.";
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp:1861]: This is no longer singular.  Not sure we want to print the entire collection.  Note that this code never executes.

File: src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp [hunk 2]
```diff
@@ -1984,28 +1907,28 @@ void RootQueueEngine::afterAppIdUnregistered(
         d_queueState_p->queue()));
 
     if (!d_isFanout) {
-        BALL_LOG_ERROR << "It should be not possible to unregister appId '"
-                       << appIdKeyPair.first << "', appKey '"
-                       << appIdKeyPair.second << "' for a non-Fanout queue.";
+        BALL_LOG_ERROR << "Invalid queue type for unregistering appId.";
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp:1910]: Same comment

File: src/groups/mqb/mqbblp/mqbblp_cluster.cpp [hunk 3]
```diff
@@ -2848,6 +2848,7 @@ void Cluster::onDomainReconfigured(const mqbi::Domain&     domain,
                                                   oldCfgAppIds,
                                                   newCfgAppIds);
 
+    // TODO: This should be one call - one QueueUpdateAdvisory for all Apps
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbblp/mqbblp_cluster.cpp:2851]: In some subsequent PR.  Not the top priority.

**Comment by kaikulimu** [comment]
Added thought, I think we should never generate appKey at https://github.com/bloomberg/blazingmq/blob/613c5baaa94f2f51b0e6575b29d96a057730b30b/src/groups/mqb/mqbc/mqbc_storageutil.cpp#L2444-L2475

Be courageous and not be afraid of changing existing code :)

File: src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp [hunk 4]
```diff
@@ -438,29 +442,33 @@ int RootQueueEngine::initializeAppId(const bsl::string& appId,
...
                          << d_queueState_p->queue()->description() << "'";
 
+        // TODO: handle w/o asserting
         BSLS_ASSERT_SAFE(false && "Virtual storage does not exist for appId");
         return -1;  // RETURN
     }
-    BSLS_ASSERT_SAFE(!appKey.isNull());
 
-    iter = makeSubStream(appId, AppKeyCount(appKey, 0), upstreamSubQueueId);
+    iter = makeSubStream(appId, appKey, upstreamSubQueueId);
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp:450]: The Dynamic Apps feature is immediate.  Over there, we do need to discern Dynamic Apps from domain-configured Apps. `initializeAppId` is one way of doing that.  Then, we need to create the `AppState` object and store the property (statically configured vs dynamically configured) in the object.
For this reason, let's revisit `initializeAppId` and `makeSubStream` after Dynamic Apps introduction.

File: src/groups/mqb/mqbblp/mqbblp_queue.cpp [hunk 5]
```diff
@@ -343,7 +343,7 @@ void Queue::convertToLocalDispatched()
 
     d_state.setId(bmqp::QueueId::k_PRIMARY_QUEUE_ID);
     createLocal();
-    rc = d_localQueue_mp->configure(errorDescription, true);
+    rc = d_localQueue_mp->configure(errorDescription, false);
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbblp/mqbblp_queue.cpp:346]: It is not reconfiguration for the newly created engine.  `true` or `false` did not matter before in this case.  Does matter after queue engine becomes aware of the `isReconfigure` value.

File: src/groups/mqb/mqbc/mqbc_clusterutil.h [hunk 6]
```diff
@@ -416,6 +416,10 @@ struct ClusterUtil {
     static void parseQueueInfo(mqbi::ClusterStateManager::AppInfos* out,
                                const bmqp_ctrlmsg::QueueInfo&       queueInfo,
                                bslma::Allocator*                    allocator);
+    static void
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbc/mqbc_clusterutil.h:419]: Both versions are being used

File: src/integration-tests/test_appids.py [hunk 7]
```diff
@@ -537,3 +537,112 @@ def test_two_consumers_of_unauthorized_app(multi_node: Cluster):
     # shutdown and wait
 
     leader.stop()
+
+
+def set_app_ids(cluster: Cluster, app_ids: List[str]):  # noqa: F811
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/integration-tests/test_appids.py:542]: I guess, NO Quality Assurance.  But I am not sure, what tool inserted.  Removing

File: src/integration-tests/test_appids.py [hunk 8]
```diff
@@ -537,3 +537,112 @@ def test_two_consumers_of_unauthorized_app(multi_node: Cluster):
...
+
+    cluster.stop_nodes()
+
+    # Reconfigure the cluster from non-FSM to FSM mode
+    for broker in cluster.configurator.brokers.values():
+        my_clusters = broker.clusters.my_clusters
+        if len(my_clusters) > 0:
+            my_clusters[0].cluster_attributes.is_cslmode_enabled = True
+            my_clusters[0].cluster_attributes.is_fsmworkflow = True
+    cluster.deploy_domains()
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/integration-tests/test_appids.py:614]: I agree, but let's defer it to another PR.  Another way, is to move this test to the `test_restart.py`

**Comment by dorjesinpo** [comment]
> Added thought, I think we should never generate appKey at
We have been over it at the previous commit when we have added ` || !appIdKeyPairs.empty()`

File: src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaikulimu) [reply src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp:1859]: Just printing the queue uri would be nice.

File: src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kaikulimu) [reply src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp:1908]: Just printing the queue uri would be nice.

File: src/groups/mqb/mqbi/mqbi_queueengine.cpp [hunk 9]
```diff
@@ -36,13 +36,13 @@ QueueEngine::~QueueEngine()
...
 void QueueEngine::afterAppIdRegistered(
-    BSLS_ANNOTATION_UNUSED const mqbi::Storage::AppInfo& appIdKeyPair)
+    BSLS_ANNOTATION_UNUSED const mqbi::Storage::AppInfos& appIdKeyPairs)
 {
     // NOTHING
 }
 
 void QueueEngine::afterAppIdUnregistered(
-    BSLS_ANNOTATION_UNUSED const mqbi::Storage::AppInfo& appIdKeyPair)
+    BSLS_ANNOTATION_UNUSED const mqbi::Storage::AppInfos& appIdKeyPairs)
```
[the remark this answers is not in the document]
Comment (kaikulimu) [reply src/groups/mqb/mqbi/mqbi_queueengine.cpp:45]: Still needs to be changed in this `.cpp` file.

File: src/groups/mqb/mqbi/mqbi_queueengine.cpp [hunk 10]
```diff
@@ -36,13 +36,13 @@ QueueEngine::~QueueEngine()
 }
 
 void QueueEngine::afterAppIdRegistered(
-    BSLS_ANNOTATION_UNUSED const mqbi::Storage::AppInfo& appIdKeyPair)
+    BSLS_ANNOTATION_UNUSED const mqbi::Storage::AppInfos& appIdKeyPairs)
```
[the remark this answers is not in the document]
Comment (kaikulimu) [reply src/groups/mqb/mqbi/mqbi_queueengine.cpp:39]: Still needs to be changed in this `.cpp` file.

**Review by kaikulimu — COMMENTED** [review]
Minor comments

**Pull request merged**